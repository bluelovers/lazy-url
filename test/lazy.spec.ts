import LazyURL from '../lib/index';

describe(`lazy`, () =>
{

	test(`lazy`, () =>
	{
		let actual = new LazyURL('docs/Web/JavaScript/Reference/Global_Objects/parseInt?file=333#7777?file=222', 'https://user:pass@developer.mozilla.org:999/zh-TW/');

		let u = new LazyURL(['docs/Web/JavaScript/Reference/Global_Objects/parseInt?file=333#7777?file=222', 'https://user:pass@developer.mozilla.org:999/zh-TW/']);

		expect(actual).toStrictEqual(u);

		expect(actual).toMatchSnapshot();
		expect(actual.toObject()).toMatchSnapshot();
		expect(actual.toString()).toMatchSnapshot();
		expect(actual.toRealString()).toMatchSnapshot();

		expect(actual.scheme).toStrictEqual(actual.protocol);
		expect(actual.scheme).toStrictEqual("https:");

		expect(actual.fragment).toStrictEqual(actual.hash);
		expect(actual.fragment).toStrictEqual('#7777?file=222');

		expect(actual.query).toStrictEqual(actual.search);
		expect(actual.query).toStrictEqual("?file=333");

		expect(actual.pathname).toStrictEqual("/zh-TW/docs/Web/JavaScript/Reference/Global_Objects/parseInt");

		expect(actual.port).toStrictEqual("999");

		expect(actual.origin).toStrictEqual("https://developer.mozilla.org:999");

		expect(actual.host).toStrictEqual("developer.mozilla.org:999");
		expect(actual.hostname).toStrictEqual("developer.mozilla.org");

	});

	test(`blob:`, () =>
	{
		let actual = new LazyURL('blob:https://user:pass@developer.mozilla.org:999/zh-TW/docs/Web/JavaScript/Reference/Global_Objects/parseInt?file=333#7777?file=222');

		expect(actual.origin).toStrictEqual("https://developer.mozilla.org:999");

		expect(actual.scheme).toStrictEqual("blob:");

		expect(actual).toMatchSnapshot();
		expect(actual.toObject()).toMatchSnapshot();
		expect(actual.toString()).toMatchSnapshot();
		expect(actual.toRealString()).toMatchSnapshot();
	})

	test(`check`, () =>
	{
		let href = 'blob:https://user:pass@developer.mozilla.org:999/zh-TW/docs/Web/JavaScript/Reference/Global_Objects/parseInt?file=333#7777?file=222';

		let u = new LazyURL(href);
		let actual = new LazyURL(u);

		expect(actual).toStrictEqual(u);
		expect(new LazyURL(new URL(href))).toStrictEqual(u);
		expect(new LazyURL({ href })).toStrictEqual(u);
		expect(new LazyURL({ href }, '')).toStrictEqual(u);
		expect(new LazyURL('', { href: actual.pathname }).toString()).toStrictEqual(actual.pathname);

		expect(() => new LazyURL(null)).toThrowErrorMatchingSnapshot();
		expect(() => new LazyURL(1 as any)).toThrowErrorMatchingSnapshot();
		expect(() => new LazyURL('')).toThrowErrorMatchingSnapshot();
		expect(() => new LazyURL(void 0)).toThrowErrorMatchingSnapshot();
		expect(() => new LazyURL(void 0, href)).toThrowErrorMatchingSnapshot();
		expect(() => new LazyURL(null, href)).toThrowErrorMatchingSnapshot();
	})

})

describe(`hostname`, () =>
{
	test(`fake`, () =>
	{
		let actual = new LazyURL('zh-TW/scripts');

		expect(actual.fakeExists()).toBeTruthy();

		expect(actual.fakeEntries()).toContainEqual(["hostname", "url-fake-hostname"]);

		actual.hostname = 'developer.mozilla.org';

		expect(actual.fakeEntries()).not.toContainEqual('hostname');

		expect(actual.toString()).toContain('developer.mozilla.org');
	});
})

describe(`port`, () =>
{

	test(`reset`, () =>
	{
		let actual = new LazyURL('https://gitee.com:3000/api/v5/');

		_checkPort(actual, '3000');

		_checkPortSet(actual, 456);

		_checkPortSet(actual, null, '');

		_checkPortSet(actual, 456);

		_checkPortSet(actual, void 0, '');

		_checkPortSet(actual, 456);

		_checkPortSet(actual, '');
	})

	test(`zero`, () =>
	{
		let actual = new LazyURL('https://gitee.com:0/api/v5/');

		_checkPort(actual, '0');

		_checkPortSet(actual, 456);
		_checkPortSet(actual, 0);

		_checkPortSet(actual, 456);
		_checkPortSet(actual, '0');
	})

	test(`number`, () =>
	{
		let actual = new LazyURL('zh-TW/scripts');

		_checkPort(actual, '');

		_checkPortSet(actual, '123');
		_checkPortSet(actual, 456);
	});

	test(`invalid`, () =>
	{
		let actual = new LazyURL('zh-TW/scripts');

		expect(() => actual.port = -1).toThrowErrorMatchingSnapshot()
		expect(() => actual.port = '-1').toThrowErrorMatchingSnapshot()
		expect(() => actual.port = Number.NaN).toThrowErrorMatchingSnapshot()
		expect(() => actual.port = Number.MAX_SAFE_INTEGER).toThrowErrorMatchingSnapshot()
		expect(() => actual.port = Infinity).toThrowErrorMatchingSnapshot()
		expect(() => actual.port = -Infinity).toThrowErrorMatchingSnapshot()
		expect(() => actual.port = ' 0xF').toThrowErrorMatchingSnapshot()
		expect(() => actual.port = ' 0123').toThrowErrorMatchingSnapshot()

	});

})

function _checkPort(actual: LazyURL, port: string)
{
	expect(actual.port).toStrictEqual(port);
	expect(actual.toString()).toContain(`:${port}/`);
}

function _checkPortSet(actual: LazyURL, port: string | number, expected?: string)
{
	actual.port = port;
	_checkPort(actual, (expected ?? port).toString())
}
