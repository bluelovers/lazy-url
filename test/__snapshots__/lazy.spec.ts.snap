// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lazy blob: 1`] = `"blob:https://user:pass@developer.mozilla.org:999/zh-TW/docs/Web/JavaScript/Reference/Global_Objects/parseInt?file=333#7777?file=222"`;

exports[`lazy blob: 2`] = `
Object {
  "hash": "#7777?file=222",
  "host": "",
  "hostname": "",
  "href": "blob:https://user:pass@developer.mozilla.org:999/zh-TW/docs/Web/JavaScript/Reference/Global_Objects/parseInt?file=333#7777?file=222",
  "password": "",
  "pathname": "https://user:pass@developer.mozilla.org:999/zh-TW/docs/Web/JavaScript/Reference/Global_Objects/parseInt",
  "port": "",
  "protocol": "blob:",
  "search": "?file=333",
  "searchParams": URLSearchParams {
    Symbol(query): Array [
      "file",
      "333",
    ],
    Symbol(context): null,
  },
  "username": "",
}
`;

exports[`lazy blob: 3`] = `"blob:https://user:pass@developer.mozilla.org:999/zh-TW/docs/Web/JavaScript/Reference/Global_Objects/parseInt?file=333#7777?file=222"`;

exports[`lazy blob: 4`] = `"blob:https://user:pass@developer.mozilla.org:999/zh-TW/docs/Web/JavaScript/Reference/Global_Objects/parseInt?file=333#7777?file=222"`;

exports[`lazy check 1`] = `"Argument 'null' is not assignable to url like."`;

exports[`lazy check 2`] = `"Argument '1' is not assignable to url like."`;

exports[`lazy check 3`] = `
"Invalid URL
    TypeError [ERR_INVALID_URL]: Invalid URL
    { input: '', baseURL: undefined }"
`;

exports[`lazy check 4`] = `"Argument 'undefined' is not assignable to url like."`;

exports[`lazy check 5`] = `"Argument 'undefined' is not assignable to url like."`;

exports[`lazy check 6`] = `"Argument 'null' is not assignable to url like."`;

exports[`lazy lazy 1`] = `"https://user:pass@developer.mozilla.org:999/zh-TW/docs/Web/JavaScript/Reference/Global_Objects/parseInt?file=333#7777?file=222"`;

exports[`lazy lazy 2`] = `
Object {
  "hash": "#7777?file=222",
  "host": "developer.mozilla.org:999",
  "hostname": "developer.mozilla.org",
  "href": "https://user:pass@developer.mozilla.org:999/zh-TW/docs/Web/JavaScript/Reference/Global_Objects/parseInt?file=333#7777?file=222",
  "password": "pass",
  "pathname": "/zh-TW/docs/Web/JavaScript/Reference/Global_Objects/parseInt",
  "port": "999",
  "protocol": "https:",
  "search": "?file=333",
  "searchParams": URLSearchParams {
    Symbol(query): Array [
      "file",
      "333",
    ],
    Symbol(context): null,
  },
  "username": "user",
}
`;

exports[`lazy lazy 3`] = `"https://user:pass@developer.mozilla.org:999/zh-TW/docs/Web/JavaScript/Reference/Global_Objects/parseInt?file=333#7777?file=222"`;

exports[`lazy lazy 4`] = `"https://user:pass@developer.mozilla.org:999/zh-TW/docs/Web/JavaScript/Reference/Global_Objects/parseInt?file=333#7777?file=222"`;

exports[`port invalid 1`] = `"Invalid port range: -1"`;

exports[`port invalid 2`] = `"Invalid port range: -1"`;

exports[`port invalid 3`] = `"Invalid port range: NaN"`;

exports[`port invalid 4`] = `"Invalid port range: 9007199254740991"`;

exports[`port invalid 5`] = `"Invalid port range: Infinity"`;

exports[`port invalid 6`] = `"Invalid port range: -Infinity"`;

exports[`port invalid 7`] = `"Invalid port input: { '0xF' => 15 }"`;

exports[`port invalid 8`] = `"Invalid port input: { '0123' => 123 }"`;
